Pass pod_spec_options to the step.

LiangquanLi930 opened the pull request:
Currently, when we use the Gangway API with the following structure:
```
"pod_spec_options": {
  "envs": {
    "xxxx": "xxxx"
  }
}
```
The variables are not being received in the step. However, in some cases, such as with the Hypershift team, this is required. For example, when performing multi-version testing, we need to change the MCE version via the Gangway API. Therefore, I'm submitting this PR to support passing pod_spec_options to the step.

psalajova commented:
Hi, can you please add a description of what is the context and reason for this change?

LiangquanLi930 commented:
> Hi, can you please add a description of what is the context and reason for this change?

Thanks for your reminder, it's been updated.

LiangquanLi930 commented on pkg/steps/pod.go:
```diff
@@ -217,6 +218,14 @@ func GenerateBasePod(
 		return nil, err
 	}
+	var pj v1.ProwJob
+	if err = client.Get(CleanupCtx, ctrlruntimeclient.ObjectKey{Namespace: jobSpec.Namespace(), Name: jobSpec.ProwJobID}, &pj); err == nil {
```
I'm not sure if the namespec is written correctly, or if we should write it as 
```
Namespace: "ci"
```
I noticed that the ns in the prowjob CRs in the job are usually "ci". For example, see this link: [prowjob.json](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_release/61690/rehearse-61690-periodic-ci-openshift-hypershift-release-4.16-periodics-mce-e2e-aws-ovn-conformance/1891033971593056256/prowjob.json). Or maybe it should be written in another way?

danilo-gemoli commented:
/hold
we are discussing [here](https://redhat-internal.slack.com/archives/CBN38N3MW/p1739802685433299?thread_ts=1739544357.642539&cid=CBN38N3MW)

LiangquanLi930 commented:
/retest

jmguzik commented on cmd/ci-operator/main.go:
```diff
@@ -786,6 +789,34 @@ func overrideMultiStageParams(o *options) error {
 
+// overrideENVListParams updates the environment variables in the test configurations based on the OVERRIDE_ENV_LIST configuration.
+// If the configuration exists, it splits the values by commas and adds the corresponding environment variables to each test configuration.
+func overrideENVListParams(o *options) {
+	viper.AutomaticEnv()
+	overrideEnvList := viper.GetString("OVERRIDE_ENV_LIST")
+	if overrideEnvList == "" {
+		return
+	}
+
+	envs := strings.Split(overrideEnvList, ",")
+
+	for _, env := range envs {
+		env = strings.TrimSpace(env)
+		envValue := viper.GetString(env)
+
+		for _, test := range o.configSpec.Tests {
+			if test.MultiStageTestConfigurationLiteral != nil {
+				if test.MultiStageTestConfigurationLiteral.Environment == nil {
+					test.MultiStageTestConfigurationLiteral.Environment = make(api.TestEnvironment)
+				}
+
+				test.MultiStageTestConfigurationLiteral.Environment[env] = envValue
+			}
+		}
+
+	}
+}
```
```suggestion
func applyEnvOverrides(o *options) {
	viper.AutomaticEnv()

	for _, envVar := range os.Environ() {
		parts := strings.SplitN(envVar, "=", 2)
		if len(parts) != 2 {
			continue
		}
		key, value := parts[0], parts[1]

		if !strings.HasPrefix(key, "OVERRIDE_") {
			continue
		}

		for _, test := range o.configSpec.Tests {
			if test.MultiStageTestConfigurationLiteral != nil {
				if test.MultiStageTestConfigurationLiteral.Environment == nil {
					test.MultiStageTestConfigurationLiteral.Environment = make(api.TestEnvironment)
				}
				test.MultiStageTestConfigurationLiteral.Environment[key] = value
			}
		}
	}
}
```

jmguzik replied:
I would change the logic of this function to pass everything starting with OVERRIDE_ prefix to multistage test.

jmguzik reviewed (changes requested):
This is added logic and I would be in favor of testing it, although it's not public interface. One way would be to do viper outside of function and later take it as param. That way function could be UTested.

jmguzik commented:
/cc @danilo-gemoli 

Danilo, PTAL as well.

LiangquanLi930 commented:
> This is added logic and I would be in favor of testing it, although it's not public interface. One way would be to do viper outside of function and later take it as param. That way function could be UTested.

Thanks for your review. I have made the adjustments, removed the viper, and added the UTest.

LiangquanLi930 commented on cmd/ci-operator/main.go:
```diff
@@ -786,6 +788,29 @@ func overrideMultiStageParams(o *options) error {
+				if test.MultiStageTestConfigurationLiteral.Environment == nil {
+					test.MultiStageTestConfigurationLiteral.Environment = make(api.TestEnvironment)
+				}
+				test.MultiStageTestConfigurationLiteral.Environment[key] = value
```
Theoretically, this is the case. We added the if just to ensure that no issues occur, in order to prevent errors when `applyEnvOverrides()` is moved in the future.

LiangquanLi930 commented:
/test e2e

danilo-gemoli commented on cmd/ci-operator/main.go:
```diff
@@ -712,6 +712,8 @@ func (o *options) Complete() error {
 		return err
 	}
 
+	applyEnvOverrides(o)
```
This should be placed before:
```golang
if err := overrideMultiStageParams(o); err != nil {
    return err
}
```
I guess cmd arguments take precedence over the env. variables.
One drawback of keeping `overrideMultiStageParams` and `applyEnvOverrides` separate is that now is quite hard to write a unit test to makes sure that cmd args override env. variables.
Unfortunately `func (o *options) Complete() error` is quite a mess to test.

LiangquanLi930 replied:
I adjusted the order of applyEnvOverrides and handleTargetAdditionalSuffix yesterday, but the E2E tests failed, so I reverted the order of these two functions.

danilo-gemoli commented on cmd/ci-operator/main.go:
```diff
@@ -786,6 +788,29 @@ func overrideMultiStageParams(o *options) error {
 	return nil
 }
 
+// applyEnvOverrides processes environment variables with override prefixes and applies them to the test configurations.
```
Comment is nice but a bit outdated:
```golang
It checks for environment variables that start with "OVERRIDE_"
```
not the case anymore, it's `MULTISTAGE_PARAM_OVERRIDE_`

danilo-gemoli commented on cmd/ci-operator/main.go:
```diff
@@ -786,6 +788,29 @@ func overrideMultiStageParams(o *options) error {
+			continue
+		}
+		key, value := parts[0], parts[1]
+		if !strings.HasPrefix(key, "MULTISTAGE_PARAM_OVERRIDE_") {
```
Move `strings.HasPrefix` at the top of this loop, before `strings.SplitN`

LiangquanLi930 replied:
An error occurs when the value of an environment variable contains MULTISTAGE_PARAM_OVERRIDE_ but the key does not contain MULTISTAGE_PARAM_OVERRIDE_.

danilo-gemoli commented on cmd/ci-operator/main_test.go:
```diff
@@ -1308,6 +1309,125 @@ func TestMultiStageParams(t *testing.T) {
+			for _, envVar := range tc.envVars {
+				parts := strings.SplitN(envVar, "=", 2)
+				if len(parts) == 2 {
+					_ = os.Setenv(parts[0], parts[1])
```
Please use `t.Setenv` rather than then `os` package. The the `defer` clause shouldn't be needed anymore.

danilo-gemoli commented on cmd/ci-operator/main_test.go:
```diff
@@ -1308,6 +1309,125 @@ func TestMultiStageParams(t *testing.T) {
+func TestApplyEnvOverrides(t *testing.T) {
+	testCases := []struct {
+		id             string
+		envVars        []string
```
Turn `envVars` into a `map[string]string` and then you can just:
```golang
for k, v := range tc.envVars {
    t.Setevn(k, v)
}
```

danilo-gemoli commented on cmd/ci-operator/main_test.go:
```diff
@@ -1308,6 +1309,125 @@ func TestMultiStageParams(t *testing.T) {
+
+			actualParams := make(map[string]string)
+
+			for _, test := range o.configSpec.Tests {
```
I would rather introduce a the field `wantConfig api.ReleaseBuildConfiguration` on the testCase struct and use that instead of extracting the env. variables we expect:
```golang
cmp.Diff(o.configSpec, tc.wantConfig)
```

LiangquanLi930 replied:
The reason I did this is that I referred to TestMultiStageParams:
https://github.com/openshift/ci-tools/blob/33124116cffe24e0877b6fce4b593b6bf30d5f88/cmd/ci-operator/main_test.go#L1204-L1309

danilo-gemoli commented:
/lgtm

danilo-gemoli commented:
/unhold

LiangquanLi930 commented:
/test integration
